### PR TITLE
Classify invalid urls

### PR DESCRIPTION
**Pull request by elsapet**
## Description
Classify URLs as invalid, potential, or valid based on their structure 
This classification is overridden in the case of a recipe match

Still pending (for a future PR)
- Reachability check

## Checklist

- [x] I've added test coverage that shows my fix or feature works as expected.
- [ ] I've updated or added documentation if required.
- [ ] I've included usage information in the description if CLI behavior was updated or added.

### CODE REVIEW

**Review by cfabianski — APPROVED** [review]
Great work!

File: pkg/util/url/url.go [hunk 1]
```diff
@@ -67,6 +241,160 @@ func PrepareURLValue(myURL string) (string, error) {
...
+	if isVendored(data.Source.Filename) {
+		ValidationResult.Reason = "included_in_vendor_folder"
+		return &ValidationResult, nil
+	}
+
+	if isPotentialDetector(data.DetectorType) {
+		ValidationResult.State = Potential
+		ValidationResult.Reason = "potential_detectors"
+		return &ValidationResult, nil
+	}
```
Comment (elsapet) [inline pkg/util/url/url.go:260]: We probably want to make these checks for all detection types (at least, it is this way in the Ruby app) - not sure the best place to put them though 🤔 
Any thoughts @vjerci @cfabianski

File: pkg/util/url/url.go [hunk 2]
```diff
@@ -93,3 +421,61 @@ func wildcardPattern(myString string, separator string) string {
 	}
 	return strings.Join(parts, regexp.QuoteMeta(separator))
 }
+
+func domainIsExcluded(host string) bool {
+	_, ok := excludedDomains[host]
+	return ok
+}
```
Comment (elsapet) [inline pkg/util/url/url.go:428]: 1. Is this sort of map/empty struct pattern okay? It was a Stack Overflow suggestion 😅 
2. Also, does it make sense in go to have something like this as a separate function, or should it all be together in the main `Validate` function?

File: pkg/util/url/url.go (hunk 1 above)
Comment (cfabianski) [reply pkg/util/url/url.go:260]: Not sure I understand the question sorry :(

File: pkg/util/url/url.go (hunk 2 above)
Comment (cfabianski) [reply pkg/util/url/url.go:428]: It felt ok when I reviewed it (for both questions) :)

File: pkg/util/url/url.go (hunk 1 above)
Comment (vjerci) [reply pkg/util/url/url.go:260]: `/util/url/` as it is at the moment looks like a good place to me

Comment (elsapet) [reply pkg/util/url/url.go:260]: It's probably me being confused :) 
When classifying detections of all type - URL, dependency, etc - we probably want to check if they are vendored or a potential detector by running the methods `isVendored` and `isPotentialDetector`. Right now these methods are private methods in the URL utils. I'm trying to ask where the best place to put them is, if we want them to be shared methods.

File: pkg/util/url/url_test.go [hunk 3]
```diff
@@ -283,3 +286,312 @@ func TestPrepareURLValue(t *testing.T) {
...
+func TestValidate(t *testing.T) {
+	tests := []struct {
+		Name       string
+		URL        string
+		Data       *report.Detection
+		isInternal bool
+		Want       *url.ValidationResult
+	}{
+		{
+			Name:       "when a detection is from a potential detector",
```
Comment (vjerci) [inline pkg/util/url/url_test.go:299]: can we not snapshot test this with coupaloy.
What i propose is to have each test case of just `URL`, `Data`, `IsInternal` and snapshot/compare results of it as a whole.

to me having 31 test cases (i suppose it is 100% code coverage) we need to manually input if this function gets feature changed feels wrong.

@cfabianski any thoughts?

**Review by vjerci — COMMENTED** [review]
I just have a question about tests which is i guess more for @cfabianski .
good job @elsapet  :)

File: pkg/util/url/url_test.go (hunk 3 above)
Comment (cfabianski) [reply pkg/util/url/url_test.go:299]: I like that way tbh. Feels easy to update/understand than a snapshot. Let’s see how it goes I guess :-)

Comment (vjerci) [reply pkg/util/url/url_test.go:299]: ok i guess thats the standard we should follow :)

File: pkg/util/url/url.go (hunk 1 above)
Comment (cfabianski) [reply pkg/util/url/url.go:260]: I'm fine with what we have now :)

**Pull request merged**